addon/v1alpha1/types_addontemplate.go

```diff
@@ -0,0 +1,181 @@
+	// HubPermissionsBindingCurrentCluster means that will only allow the addon agent to access the
+	// resources in managed cluster namespace on the hub cluster.
+	// It is a specific case of the SingleNamespace type.
+	HubPermissionsBindingCurrentCluster HubPermissionsBindingType = "CurrentCluster"
+)
+
+// RegistrationSpec describes how to register an addon agent to the hub cluster.
+// With the registration defined, The addon agent can access to kube apiserver with kube style API
+// or other endpoints on hub cluster with client certificate authentication. During the addon
+// registration process, a csr will be created for each Registration on the hub cluster. The
+// CSR will be approved automatically, After the csr is approved on the hub cluster, the klusterlet
+// agent will create a secret in the installNamespace for the addon agent.
+// If the RegistrationType type is KubeClient, the secret name will be "{addon name}-hub-kubeconfig"
+// whose content includes key/cert and kubeconfig. Otherwise, If the RegistrationType type is
+// CustomSigner the secret name will be "{addon name}-{signer name}-client-cert" whose content
+// includes key/cert.
+type RegistrationSpec struct {
+	// Type of the registration configuration, it supports:
+	// - KubeClient: the addon agent can access the hub kube apiserver with kube style API.
+	//   the signer name should be "kubernetes.io/kube-apiserver-client". When this type is
+	//   used, the KubeClientRegistrationConfig can be used to define the permission of the
+	//   addon agent to access the hub cluster
+	// - CustomSigner: the addon agent can access the hub cluster through user-defined endpoints.
+	//   When this type is used, the CustomSignerRegistrationConfig can be used to define how
+	//   to issue the client certificate for the addon agent.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=KubeClient;CustomSigner
+	Type RegistrationType `json:"type"`
+
+	// KubeClient holds the configuration of the KubeClient type registration
+	// +optional
+	KubeClient *KubeClientRegistrationConfig `json:"kubeClient,omitempty"`
+
+	// CustomSigner holds the configuration of the CustomSigner type registration
+	// required when the Type is CustomSigner
+	CustomSigner *CustomSignerRegistrationConfig `json:"customSigner,omitempty"`
+}
+
+type KubeClientRegistrationConfig struct {
+	// HubPermissions represent the permission configurations of the addon agent to access the hub cluster
+	// +optional
+	HubPermissions []HubPermissionConfig `json:"hubPermissions,omitempty"`
+}
+
+// HubPermissionConfig configures the permission of the addon agent to access the hub cluster.
+// Will create a RoleBinding in the same namespace as the managedClusterAddon to bind the user
+// provided ClusterRole/Role to the "system:open-cluster-management:cluster:<cluster-name>:addon:<addon-name>"
+// Group.
+type HubPermissionConfig struct {
+	// Type of the permissions setting. It defines how to bind the roleRef
```

Comment:
we need to describe the meaning of CurrentCluster and singlenamespace

Reply:
described [here](https://github.com/open-cluster-management-io/api/pull/239/files#diff-cc2eb9f759617220330a98f0609b906346b9f986ccaf027c4449db17fbf1f4dcR66-R72), should I move them into the `HubPermissionConfig` struct？

Reply:
I think we should for generating the openapi doc

Reply:
description added